src/dataflow/deadlines.rs

```diff
@@ -13,105 +25,78 @@ use tokio::time::Duration;
 pub trait CondFn: Fn(&ConditionContext, &Timestamp) -> bool + Send + Sync {}
 impl<F: Fn(&ConditionContext, &Timestamp) -> bool + Send + Sync> CondFn for F {}
 
-/// A `DeadlineEvent` structure defines a deadline that is generated upon the fulfillment of a
-/// start condition on a given stream and a given timestamp (we assume a single deadline for each
-/// timestamp). Upon expiration of the deadline (defined as `duration`), the `end_condition`
-/// function is checked, and the `handler` is invoked if it was not satisfied.
-pub struct DeadlineEvent {
-    pub stream_id: StreamId,
-    pub timestamp: Timestamp,
-    pub duration: Duration,
-    pub handler: Arc<Mutex<dyn HandlerContextT>>,
-    pub end_condition: Arc<dyn CondFn>,
-}
+/*************************************************************************************************
+ * DeadlineFn: A DeadlineFn defines the duration between the start and the end condition that an *
+ * operator should be restricted to.                                                             *
+ ************************************************************************************************/
 
-impl DeadlineEvent {
-    pub fn new(
-        stream_id: StreamId,
-        timestamp: Timestamp,
-        duration: Duration,
-        handler: Arc<Mutex<dyn HandlerContextT>>,
-        end_condition: Arc<dyn CondFn>,
-    ) -> Self {
-        Self {
-            stream_id,
-            timestamp,
-            duration,
-            handler,
-            end_condition,
-        }
-    }
-}
+/// A trait that defines the deadline function.
+pub trait DeadlineFn<S>: FnMut(&S, &Timestamp) -> Duration + Send + Sync {}
```

Comment:
Does it make sense to compute a duration for the deadline rather than a specific point in time at which the deadline expires (e.g. `SystemTime`)?

Reply:
Yeah, it can go either way. It just depends on what the prevalent use-case is. I believe that it is easier to define the duration of the deadline, than an exact point in time when it expires.

@ICGog can you chime in here?

Reply:
Agreed, this can go either way. I'm tempted to treat these as relative deadlines for now (i.e., compute duration) rather than absolute deadlines because this code implements timestamp deadlines.